onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc

```diff
@@ -631,25 +725,57 @@ bool BinaryOpBuilder::IsOpSupportedImpl(ModelBuilder& /* model_builder */, const
     return false;
   }
 
+  if (op_is_qlinear) {
+    // For QLinearAdd, we only support uint8 output now
+    int32_t output_type;
+    if (!GetType(*node.OutputDefs()[0], output_type))
+      return false;
+
+    if (output_type != ONNX_NAMESPACE::TensorProto_DataType_UINT8) {
+      LOGS_DEFAULT(VERBOSE) << "[" << op_type
+                            << "] output type: [" << output_type
+                            << "] is not supported for now";
+      return false;
+    }
+
+    // All scale/zero points are initializer scalars
+    // a/b/y_scale
+    if (!IsQuantizationScaleSupported(model_builder, node, {1, 4, 6}))
+      return false;
+
+    // a/b/y_zero_point
+    if (!IsQuantizationZeroPointSupported(model_builder, node, {2, 5, 7}))
+      return false;
+  }
+
   return true;
 }
 
 void BinaryOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node) {
-  const auto& op(node.OpType());
+  const auto& op_type(node.OpType());
+  const auto input_defs(node.InputDefs());
+
   int32_t op_code;
-  if (op == "Add")
+  bool op_is_qlinear = op_type == "QLinearAdd";
+  if (op_type == "Add" || op_is_qlinear)
     op_code = ANEURALNETWORKS_ADD;
-  else if (op == "Sub")
+  else if (op_type == "Sub")
     op_code = ANEURALNETWORKS_SUB;
-  else if (op == "Mul")
+  else if (op_type == "Mul")
     op_code = ANEURALNETWORKS_MUL;
-  else if (op == "Div")
+  else if (op_type == "Div")
     op_code = ANEURALNETWORKS_DIV;
   else {
-    ORT_THROW("UnaryOpBuilder, unknown op: " + op);
+    ORT_THROW("UnaryOpBuilder, unknown op: " + op_type);
+  }
+
+  size_t a_idx = 0, b_idx = 1;
+  if (op_is_qlinear) {
+    b_idx = 3;
   }
-  std::string input1 = node.InputDefs()[0]->Name();
-  std::string input2 = node.InputDefs()[1]->Name();
+
+  std::string input1 = input_defs[a_idx]->Name();
+  std::string input2 = input_defs[b_idx]->Name();
   const auto& output = node.OutputDefs()[0]->Name();
```

Comment:
const std::string&

Reply:
These input names may change in the code below (line 791,798) so I am copying the name here

---
In reply to: [462081004](https://github.com/microsoft/onnxruntime/pull/4607#discussion_r462081004) [](ancestors = 462081004)